api/include/opentelemetry/context/context.h

```diff
@@ -0,0 +1,244 @@
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace context
+{
+
+  // The context class provides a context identifier. Is built as a linked list
+  // of context nodes which each have access to the nodes before it, but not 
+  // after. So each time a write is performed a new context is added to the list.
+  // Within each context node there is another linked list which holds the key
+  // and value data for each node.
+  class Context
+  {
+
+    public:
+
+
+      Context() = default;
+
+      // Creates a context object from a map of keys and identifiers, this will
+      // be the head of the context object linked list.
+      template <class T , nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+        Context(const T &keys_and_values)  
+        {
+          head_ = new DataList(keys_and_values);
+        }
+
+      // Creates a context object from a key and value, this will be the head
+      // of the context object linked list.
+      Context(nostd::string_view key, ContextValue value)  
+      {
+        head_ = new DataList(key, value);
+      }
+
+      // Accepts a new iterable and then returns a new context that
+      // contains the new key and value data.
+      template <class T , nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+        Context* SetValues(T &values) noexcept
+        {
+          return new Context(values, this);
+        }
+
+      // Accepts a new iterable and then returns a new context that
+      // contains the new key and value data.
+      Context* SetValue(nostd::string_view key, ContextValue value) noexcept
+      {
+        return new Context(key, value, this);
+      }
+
+      // Returns the value associated with the passed in key.
+      common::AttributeValue GetValue(nostd::string_view key)
+      {
+        //Iterate through the context nodes
+        for(Context* context = this; context != nullptr; context = context->next_){
+          //Iterate through the internal data nodes
+          for(DataList* data = context->head_; data != nullptr; data = data->next_){
+            if(key.length() == data->key_length_ && strncmp(key.data(), data->key_, data->key_length_) == 0){
+              return data->value_;
+            }
+          }
+        }
+        return "";
+      }
+
+      // Iterates through the current and comparing context objects
+      // to check for equality.
+      bool operator==(const Context &other)
+      {
+        // If the two contexts are not at the place in the list
+        // they cannot be equivalent
+        if(next_ != other.next_){
+          return false; 
+        }
+
+        DataList* other_data = other.head_;
+
+        // Iterate through the lists and compare the keys
+        // TODO should also compare the values but the nostd::variant
+        // seems to not have a comparison operator
+        for(DataList* data = head_; data != nullptr; data = data->next_){
```

Comment:
Let's rather implement a dedicated comparator on `DataList`. Also, besides not taking values into account, the code here assumes that the keys are ordered in the same way.

Reply:
When you are comparing two contexts what exactly are you checking for? I was expecting that two contexts with differently ordered keys would be considered different contexts. Also I have been having problems with nostd::variant comparison, I am planning on looking into it after getting the context dealt with.